build.zig

```diff
@@ -0,0 +1,1484 @@
+    b: *Build,
+
+    fn init(b: *Build) CSources {
+        return .{
+            .files = std.ArrayList([]const u8).init(b.allocator),
+            .b = b,
+        };
+    }
+
+    fn collect(sources: *CSources, path: []const u8) !void {
+        const cwd: std.fs.Dir = sources.b.build_root.handle;
+        const dir: std.fs.Dir = try cwd.openDir(path, .{ .iterate = true });
+        var iter: std.fs.Dir.Iterator = dir.iterate();
+        while (try iter.next()) |entry| {
+            if (entry.kind == .file and std.mem.endsWith(u8, entry.name, ".c")) {
+                const filepath = try std.fs.path.resolve(sources.b.allocator, &.{ path, entry.name });
+                try sources.files.append(filepath);
+            }
+        }
+    }
+
+    fn deinit(sources: CSources) void {
+        for (sources.files.items) |path| {
+            sources.b.allocator.free(path);
+        }
+        sources.files.deinit();
+    }
+};
+
+fn makeDir(path: []const u8) !void {
+    const cwd = std.fs.cwd();
+    cwd.makeDir(path) catch |e| {
+        if (e != error.PathAlreadyExists) {
+            return e;
+        }
+    };
+}
+
+fn pathExists(dir: std.fs.Dir, path: []const u8) bool {
+    dir.access(path, .{}) catch {
+        return false;
+    };
+    return true;
+}
+
+const LibShas = struct {
+    angle: []const u8,
+    dawn: []const u8,
+
+    fn find(cwd: std.fs.Dir, allocator: std.mem.Allocator) !LibShas {
+        const Helpers = struct {
+            fn findFieldIndex(ast: *const std.zig.Ast, current_node: u32, search: []const u8) ?u32 {
+                var buf: [2]std.zig.Ast.Node.Index = undefined;
+                if (ast.fullStructInit(&buf, current_node)) |struct_ast| {
+                    for (struct_ast.ast.fields) |i| {
+                        const field_name = ast.tokenSlice(ast.firstToken(i) - 2);
+                        if (std.mem.eql(u8, field_name, search)) {
+                            return i;
+                        }
+                    }
+                }
+                return null;
+            }
+
+            // "https://chromium.googlesource.com/angle/angle.git/+archive/8a8c8fc280d74b34731e0e417b19bff7c967388a.tar.gz"
+            fn extractCommitFromUrl(url: []const u8) []const u8 {
+                const basename = std.fs.path.basename;
+                const stem = std.fs.path.stem;
+                return stem(stem(basename(url)));
+            }
+        };
+
+        const contents: []const u8 = try cwd.readFileAlloc(allocator, "build.zig.zon", 1024 * 1024 * 1);
+        const contents_z = try allocator.alloc(u8, contents.len + 1);
+        @memcpy(contents_z[0..contents.len], contents);
+        contents_z[contents.len] = 0;
+
+        // NOTE: there will eventually be a ZON parser in the stdlib, but until then we'll just get by
+        // with a hacky version that only looks for the libs we care about
+        var ast = try std.zig.Ast.parse(allocator, contents_z[0..contents.len :0], .zon);
```

Comment:
@Parzival-3141 mentioned we should be able to use `std.zon` to parse this now instead of this approach.